### ISSUE TITLE
Incorrectly inferred array type when merging in loop

### ISSUE THREAD

**Issue by AndrolGenhald** [vimeo/psalm#7537]
All of these incorrectly infer `$array` to have exactly two values:
https://psalm.dev/r/2cdcc77b29
https://psalm.dev/r/6ce02613d5 (also incorrectly has `0` key possibly undefined)
https://psalm.dev/r/6c18545837 (also incorrectly has `0` and `1` keys possibly undefined)
https://psalm.dev/r/6aefa34537 (also incorrectly has `0` key possibly undefined)

**Comment by orklah** [comment]
@weirdan seems like the bot is dead :(

**Comment by weirdan** [comment]
It wouldn't post any comments if it was dead. It's psalm.dev API that returns 500 Internal server error: https://psalm.dev/r/2cdcc77b29/results

**Comment by weirdan** [comment]
@muglug do we have any access to the logs of psalm.dev server?

**Comment by muglug** [comment]
Here's what I see. It's intermittent.

```
Got error 'PHP message: PHP Fatal error: Uncaught TypeError: {closure}(): Argument #1 ($val) must be of type string, null given in /var/www/vhosts/psalm.dev/httpdocs/views/snippet.php:70
Stack trace:\n#0 [internal function]: {closure}()
#1 /var/www/vhosts/psalm.dev/httpdocs/views/snippet.php(72): array_map()
#2 {main}\n thrown in /var/www/vhosts/psalm.dev/httpdocs/views/snippet.php on line 70
```

**Comment by orklah** [comment]
Mmh, I wonder if it couldn't be a limitation of the API. I heavily used the tool to check for resolved issue. Maybe it hit a quota?

**Comment by muglug** [comment]
I think it's a MySQL issue — I'm doing a backfill atm and the errors all come at once

**Comment by weirdan** [comment]
@muglug I see you altered the codes.sql rather than adding a migration?

**Comment by weirdan** [comment]
Also mysql docker container starts with empty database:
```
MySQL psalm_mysql_user@127.0.0.1:psalm_web> show tables;
╒═════════════════════╕
│ Tables_in_psalm_web │
╞═════════════════════╡
╘═════════════════════╛
```

**Comment by weirdan** [comment]
@muglug do you plan to alter those columns you added to `NOT NULL` after the backfill is complete?

**Comment by muglug** [comment]
Let's ticket move the discussion here: https://github.com/psalm/psalm.dev/issues/71

**Comment by weirdan** [comment]
psalm.dev/r/2cdcc77b29
psalm.dev/r/6ce02613d5
psalm.dev/r/6c18545837
psalm.dev/r/6aefa34537

**Comment by davidbyoung** [comment]
Is this related to what I am seeing here (not in a loop):  https://psalm.dev/r/f1b6a3136e

**Comment by AndrolGenhald** [comment]
@davidbyoung That looks like #7297.

**Comment by davidbyoung** [comment]
Ah, thanks! Tried searching, I should have tried harder haha

**Comment by orklah** [comment]
@AndrolGenhald array have been revamped completely. Mind taking a look if there's still errors?

**Comment by danog** [comment]
Yeah this is a simple case of some old psalm code making use of LNumber instead of using the type provider to extract the statement type, that's a simple fix in theory
